actions/lib/purge_remote_folder.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/python
+
+import argparse
+import datetime
+import json
+import logging
+import os
+import shutil
+import socket
+
+
+class PurgeRemoteFolders(object):
+
+    def __init__(self, biotank_archive_folder, age_in_days, dryrun):
+        self.biotank_archive_folder = str(biotank_archive_folder)
+        self.age_in_days = int(age_in_days)
+        self.dryrun = bool(dryrun)
+
+        hostname = socket.gethostname()
+        self.directory = self.sanitize_path(
+            self.biotank_archive_folder,
+            os.path.join(
+                "/data",
+                hostname))
+        self.logfile = os.path.join(self.directory, "KEEP_removed_archives.log")
+
+    @staticmethod
+    def sanitize_path(dirty_path, directory):
+        if os.path.isabs(dirty_path):
+            raise ValueError(
+                "{p} is an absolute path which is not allowed at this point".format(
+                    p=dirty_path))
+        # fully resolve path to make sure that no symlinks or upward references are left and cause surprises
+        clean_path = os.path.realpath(
+            os.path.join(directory, dirty_path))
+        # ensure that the resulting path is below the directory
+        if not clean_path.startswith(directory):
+            raise ValueError(
+                "{p} appears to resolve to a path outside of {d} which is not allowed at this point".format(
+                    p=dirty_path, d=directory))
+        return clean_path
+
+    def get_files_and_folders(self):
+        older_than = datetime.date.today() - datetime.timedelta(days=self.age_in_days)
+
+        def _path_should_be_purged(p):
+            return datetime.date.fromtimestamp(
+                os.stat(
+                    os.path.join(
+                        self.directory, p)).st_mtime) < older_than and not p.startswith("KEEP")
+
+        return filter(
+            _path_should_be_purged,
+            os.listdir(self.directory))
+
+    def purge_files_and_folders(self, files_and_folders_to_purge, logger):
+
+        def _purge_path(p):
+            purge_fn = shutil.rmtree if os.path.isdir(p) else os.unlink
+            if not self.dryrun:
+                purge_fn(p)
+                logger.info(p)
+
+        map(
```

Comment:
Not very important, but here I actually thing that a nested loop could be cleared that a nested map. As you know I am a great fan of the functional idioms, but since they are less common in Python using for loops might be preferred here.

Reply:
ok, makes sense